Fix `libparsec_protocol` not detecting a rebuild is needed when a .json5 schema is added

touilleMan opened the pull request.

FirelightFlagboy commented:
1. I do not see how that fix the issue
2. That increase the size of the generated data which slow down compilation

touilleMan commented:
> I do not see how that fix the issue

consider the following:
1. `cargo build -p libparsec_protocol`
2. `git checkout otherbranch`
3. `cargo build -p libparsec_protocol`

If `otherbranch` introduces a new schema, that means a new .json5 file is present (let's say in `libparsec/crates/protocol/schema/authenticated_cmds/new_command.json5`).
However build done at step 1 has produced `cargo:rerun-if-changed=` commands according to a `glob("schema/**/*.json5")` that occured *before* this file was present !
So in the end the build done is step 3 doesn't detect a re-build is needed and does nothing, which in turn produce rather strange errors such as (real life example ^^):
```
[INFO]: 🌀  Compiling to Wasm...
   Compiling libparsec_account v3.4.1-a.0+dev (/home/max/parsec-cloud/libparsec/crates/account)
error[E0432]: unresolved import `libparsec_protocol::authenticated_account_cmds::latest::auth_method_create`
  --> libparsec/crates/account/src/create_auth_method.rs:27:65
   |
27 |     use libparsec_protocol::authenticated_account_cmds::latest::auth_method_create::{Rep, Req};
   |                                                                 ^^^^^^^^^^^^^^^^^^ could not find `auth_method_create` in `latest`

For more information about this error, try `rustc --explain E0432`.
error: could not compile `libparsec_account` (lib) due to 1 previous error```

> That increase the size of the generated data which slow down compilation

The compiler is supposed to be smart enough to totally strip those data since they are private to the module and never used.

touilleMan commented:
I've just tested the compilation time:
With this patch
```shell
$ cargo clean -p libparsec_protocol
     Removed 277 files, 277.2MiB total

$ time cargo build -p libparsec_protocol
   Compiling libparsec_protocol v3.4.1-a.0+dev (/home/emmanuel/projects/parsec-cloud/libparsec/crates/protocol)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 5.35s

________________________________________________________
Executed in    5,23 secs    fish           external
   usr time    8,63 secs    0,00 millis    8,63 secs
   sys time    1,22 secs    1,04 millis    1,22 secs
```

Without this patch:
```shell
$ cargo clean -p libparsec_protocol
     Removed 277 files, 277.2MiB total

$ time cargo build -p libparsec_protocol
   Compiling libparsec_protocol v3.4.1-a.0+dev (/home/emmanuel/projects/parsec-cloud/libparsec/crates/protocol)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 6.51s

________________________________________________________
Executed in    5,17 secs    fish           external
   usr time    8,49 secs  143,00 micros    8,49 secs
   sys time    1,26 secs  970,00 micros    1,25 secs
```

(I've run both build multiple times and got time in the range 5.1s to 6.4s, so no significant difference seems to be introduced by this patch)

FirelightFlagboy commented:
I understand the scenario, but I don't see how to this PR solve it

touilleMan commented:
> I understand the scenario, but I don't see how to this PR solve it

Good catch ! My assumption were wrong and this "fix" doesn't fix anything :smile: 

However there is still hope: I've just discovered that `cargo:rerun-if-changed` also support directories \o/
So I've updated this PR to use this instead (and it seems to work fine now ^^)

FirelightFlagboy commented:
Can you do the same for types ?

touilleMan commented:
> I think we can remove the glob dependencies now 🤔

:+1: done !
